File: src/ai_composer/markdown_loader.py
from __future__ import annotations

from pathlib import Path
from typing import Any

from .models import ArtifactNode


def _parse_scalar(value: str) -> Any:
    value = value.strip()
    if value.lower() in {"true", "false"}:
        return value.lower() == "true"
    if value.isdigit():
        return int(value)
    if value.startswith("[") and value.endswith("]"):
        inner = value[1:-1].strip()
        if not inner:
            return []
        return [item.strip().strip("'\"") for item in inner.split(",")]
    return value.strip("'\"")


def _parse_frontmatter(frontmatter: str) -> dict[str, Any]:
    meta: dict[str, Any] = {}
    current_list_key: str | None = None
    for raw_line in frontmatter.splitlines():
        line = raw_line.rstrip()
        if not line.strip():
            continue
        stripped = line.strip()
        if stripped.startswith("- ") and current_list_key:
            meta.setdefault(current_list_key, []).append(_parse_scalar(stripped[2:]))
            continue
        if ":" not in line:
            continue
        key, value = line.split(":", 1)
        key = key.strip()
        value = value.strip()
        if not value:
            meta[key] = []
            current_list_key = key
            continue
        meta[key] = _parse_scalar(value)
        current_list_key = None
    return meta


def load_markdown_artifacts(root_dir: str) -> dict[str, ArtifactNode]:
    root = Path(root_dir)
    nodes: dict[str, ArtifactNode] = {}
    for path in sorted(root.rglob("*.md")):
        raw = path.read_text(encoding="utf-8")
        metadata: dict[str, Any] = {}
        body = raw
        if raw.startswith("---\n"):
            _, frontmatter, remainder = raw.split("---\n", 2)
            metadata = _parse_frontmatter(frontmatter)
            body = remainder.strip()
        node_id = str(metadata.get("id") or path.stem)
        node_type = str(metadata.get("type") or "document").lower()
        title = str(metadata.get("title") or path.stem.replace("-", " ").title())
        nodes[node_id] = ArtifactNode(
            id=node_id,
            type=node_type,
            title=title,
            body=body,
            metadata=metadata,
            path=str(path),
        )
    _link_nodes(nodes)
    return nodes


def _link_nodes(nodes: dict[str, ArtifactNode]) -> None:
    for node in nodes.values():
        links: list[str] = []
        if node.type == "ticket":
            epic = node.metadata.get("epic")
            if isinstance(epic, str) and epic in nodes:
                links.append(epic)
            rules = node.metadata.get("rules", [])
            if isinstance(rules, list):
                links.extend(rule for rule in rules if isinstance(rule, str) and rule in nodes)
        if node.type == "adr":
            ticket = node.metadata.get("ticket")
            if isinstance(ticket, str) and ticket in nodes:
                links.append(ticket)
        node.links = list(dict.fromkeys(links))

File: src/AiComposer.Core/Markdown/MarkdownLoader.cs
using AiComposer.Core.Models;

namespace AiComposer.Core.Markdown;

/// <summary>
/// Parses markdown files with YAML frontmatter into <see cref="ArtifactNode"/> objects
/// and assembles them into a knowledge graph by resolving inter-node links.
/// </summary>
public static class MarkdownLoader
{
    /// <summary>
    /// Recursively discovers all *.md files under <paramref name="rootDir"/>,
    /// parses their frontmatter and body, then links related nodes.
    /// </summary>
    public static Dictionary<string, ArtifactNode> LoadArtifacts(string rootDir)
    {
        var nodes = new Dictionary<string, ArtifactNode>(StringComparer.Ordinal);

        foreach (var file in Directory.EnumerateFiles(rootDir, "*.md", SearchOption.AllDirectories)
                                      .OrderBy(f => f))
        {
            var node = ParseFile(file);
            nodes[node.Id] = node;
        }

        LinkNodes(nodes);
        return nodes;
    }

    // ── private helpers ────────────────────────────────────────────────────────

    private static ArtifactNode ParseFile(string path)
    {
        var raw = File.ReadAllText(path);
        var metadata = new Dictionary<string, object>(StringComparer.OrdinalIgnoreCase);
        var body = raw;

        if (raw.StartsWith("---\n", StringComparison.Ordinal))
        {
            var parts = raw.Split(["---\n"], 3, StringSplitOptions.None);
            if (parts.Length >= 3)
            {
                metadata = ParseFrontmatter(parts[1]);
                body = parts[2].Trim();
            }
        }

        var stem = System.IO.Path.GetFileNameWithoutExtension(path);
        var id = metadata.TryGetValue("id", out var rawId) ? rawId?.ToString() ?? stem : stem;
        var type = metadata.TryGetValue("type", out var rawType)
            ? (rawType?.ToString() ?? "document").ToLowerInvariant()
            : "document";
        var title = metadata.TryGetValue("title", out var rawTitle)
            ? rawTitle?.ToString() ?? ToTitle(stem)
            : ToTitle(stem);

        return new ArtifactNode
        {
            Id = id,
            Type = type,
            Title = title,
            Body = body,
            Metadata = metadata,
            Path = path,
        };
    }

    private static Dictionary<string, object> ParseFrontmatter(string frontmatter)
    {
        var result = new Dictionary<string, object>(StringComparer.OrdinalIgnoreCase);
        string? currentListKey = null;

        foreach (var rawLine in frontmatter.Split('\n'))
        {
            var line = rawLine.TrimEnd();
            if (string.IsNullOrWhiteSpace(line)) continue;

            var stripped = line.TrimStart();

            // List item under the previous key
            if (stripped.StartsWith("- ", StringComparison.Ordinal) && currentListKey is not null)
            {
                if (!result.ContainsKey(currentListKey))
                    result[currentListKey] = new List<string>();
                ((List<string>)result[currentListKey]).Add(stripped[2..].Trim().Trim('\'', '"'));
                continue;
            }

            var colon = line.IndexOf(':', StringComparison.Ordinal);
            if (colon < 0) continue;

            var key = line[..colon].Trim();
            var value = line[(colon + 1)..].Trim();

            if (string.IsNullOrEmpty(value))
            {
                result[key] = new List<string>();
                currentListKey = key;
                continue;
            }

            result[key] = ParseScalar(value);
            currentListKey = null;
        }

        return result;
    }

    private static object ParseScalar(string value)
    {
        value = value.Trim().Trim('\'', '"');
        if (value.Equals("true", StringComparison.OrdinalIgnoreCase)) return true;
        if (value.Equals("false", StringComparison.OrdinalIgnoreCase)) return false;
        if (int.TryParse(value, out var i)) return i;

        if (value.StartsWith('[') && value.EndsWith(']'))
        {
            var inner = value[1..^1].Trim();
            if (string.IsNullOrEmpty(inner)) return new List<string>();
            return inner.Split(',').Select(s => s.Trim().Trim('\'', '"')).ToList();
        }

        return value;
    }

    private static void LinkNodes(Dictionary<string, ArtifactNode> nodes)
    {
        foreach (var node in nodes.Values)
        {
            var links = new List<string>();

            if (node.Type == "ticket")
            {
                if (node.Metadata.TryGetValue("epic", out var epic) && epic is string epicId && nodes.ContainsKey(epicId))
                    links.Add(epicId);

                if (node.Metadata.TryGetValue("rules", out var rules) && rules is List<string> ruleIds)
                    links.AddRange(ruleIds.Where(r => nodes.ContainsKey(r)));
            }

            if (node.Type == "adr")
            {
                if (node.Metadata.TryGetValue("ticket", out var ticket) && ticket is string tId && nodes.ContainsKey(tId))
                    links.Add(tId);
            }

            node.Links = links.Distinct().ToList();
        }
    }

    private static string ToTitle(string stem) =>
        System.Text.RegularExpressions.Regex.Replace(
            stem.Replace("-", " "),
            @"\b\w",
            m => m.Value.ToUpperInvariant());
}
